src/mod_presence.erl

```diff
@@ -0,0 +1,611 @@
+    ejabberd_router:route(FromJid, ToJid, Acc, Packet),
+    I = gb_sets:add_element(ToJid, Handler#presences.pres_i),
+    A = gb_sets:del_element(ToJid, Handler#presences.pres_a),
+    NewHandler = Handler#presences{pres_i = I, pres_a = A},
+    mongoose_c2s_acc:to_acc(Acc, state_mod, {?MODULE, NewHandler}).
+
+process_presence_track_subscription_and_route(Acc, FromJid, ToJid, Packet, Type) ->
+    Acc1 = mongoose_hooks:roster_out_subscription(Acc, FromJid, ToJid, Type),
+    ejabberd_router:route(jid:to_bare(FromJid), ToJid, Acc1, Packet),
+    Acc1.
+
+-spec presence_broadcast(mongoose_acc:t(), jid_set()) -> mongoose_acc:t().
+presence_broadcast(Acc, JIDSet) ->
+    From = mongoose_acc:from_jid(Acc),
+    gb_sets:fold(fun(JID, A) ->
+                         ejabberd_router:route(From, JID, A)
+                 end, Acc, JIDSet).
+
+-spec presence_update_to_available(
+        mongoose_acc:t(), jid:jid(), jid:jid(), exml:element(), mongoose_c2s:c2s_data(),
+        presences(), list(), priority(), priority(), boolean()) ->
+    mongoose_acc:t().
+presence_update_to_available(Acc, FromJid, _ToJid, Packet, StateData, Handler, SocketSend,
+                             _OldPriority, NewPriority, true) ->
+    Acc1 = mongoose_hooks:user_available_hook(Acc, FromJid),
+    Acc2 = case NewPriority >= 0 of
+              true ->
+                  resend_offline_messages(Acc1, StateData);
+              false ->
+                  Acc1
+              end,
+    presence_broadcast_first(Acc2, FromJid, Packet, Handler, SocketSend);
+presence_update_to_available(Acc, FromJid, _ToJid, Packet, StateData, Handler, Pending,
+                             OldPriority, NewPriority, false) ->
+    presence_broadcast_to_trusted(
+             Acc, FromJid, Handler#presences.pres_f, Handler#presences.pres_a, Packet),
+    Acc1 = case OldPriority < 0 andalso NewPriority >= 0 of
+               true ->
+                   resend_offline_messages(Acc, StateData);
+               false ->
+                   Acc
+           end,
+    ToAcc = [{socket_send, Pending}, {state_mod, {?MODULE, Handler}}],
+    mongoose_c2s_acc:to_acc_many(Acc1, ToAcc).
+
+-spec presence_broadcast_to_trusted(
+        mongoose_acc:t(), jid:jid(), jid_set(), jid_set(), exml:element()) ->
+    mongoose_acc:t().
+presence_broadcast_to_trusted(Acc, FromJid, T, A, Packet) ->
+    gb_sets:fold(
+      fun(JID, Ac) ->
+              case gb_sets:is_element(JID, T) of
+                  true ->
+                      ejabberd_router:route(FromJid, JID, Ac, Packet);
+                  _ ->
+                      Ac
+              end
+      end, Acc, A).
+
+-spec resend_offline_messages(mongoose_acc:t(), mongoose_c2s:c2s_data()) -> mongoose_acc:t().
+resend_offline_messages(Acc, StateData) ->
+    ?LOG_DEBUG(#{what => resend_offline_messages, acc => Acc, c2s_state => StateData}),
+    Jid = mongoose_c2s:get_jid(StateData),
+    Acc1 = mongoose_hooks:resend_offline_messages_hook(Acc, Jid),
```

Comment:
I wonder if `mod_offline` could do all the logic below by itself in the hook above instead of having this `mod_offline`-related code here.

Reply:
Good point. Maybe in a future rework of offline and/or presences, this would be good to cleanup, but it falls a bit outside of this PR 😕

Reply:
Ok we can do it later.